Cubical/Algebra/NatSolver/HornerForms.agda

```diff
@@ -0,0 +1,100 @@
+{-# OPTIONS --safe #-}
+module Cubical.Algebra.NatSolver.HornerForms where
+
+open import Cubical.Foundations.Prelude
+
+open import Cubical.Data.Nat
+open import Cubical.Data.FinData
+open import Cubical.Data.Vec
+open import Cubical.Data.Bool using (Bool; true; false; if_then_else_)
+
+private
+  variable
+    ℓ : Level
+
+{-
+  This defines the type of multivariate Polynomials over ℕ.
+  The construction is based on the algebraic fact
+
+    ℕ[X₀][X₁]⋯[Xₙ] ≅ ℕ[X₀,⋯,Xₙ]
+
+  BUT: Contrary to algebraic convetions, we will give 'Xₙ' the lowest index
+  in the definition of 'Variable' below. So if 'Variable n k' is identified
+  with 'Xₖ', what we construct should rather be denoted with
+
+    ℕ[Xₙ][Xₙ₋₁]⋯[X₀]
+
+  or, to be precise about the evaluation order:
+
+    (⋯((ℕ[Xₙ])[Xₙ₋₁])⋯)[X₀]
+
+-}
+
+data IteratedHornerForms : ℕ → Type ℓ-zero where
+  const : ℕ → IteratedHornerForms ℕ.zero
+  0H : {n : ℕ} → IteratedHornerForms (ℕ.suc n)
+  _·X+_ : {n : ℕ} → IteratedHornerForms (ℕ.suc n) → IteratedHornerForms n
+                  → IteratedHornerForms (ℕ.suc n)
+
+eval : (n : ℕ) (P : IteratedHornerForms n)
+       → Vec ℕ n → ℕ
+eval ℕ.zero (const r) [] = r
+eval .(ℕ.suc _) 0H (_ ∷ _) = 0
+eval (ℕ.suc n) (P ·X+ Q) (x ∷ xs) =
+  (eval (ℕ.suc n) P (x ∷ xs)) · x + eval n Q xs
```

Comment:
I think the `ℕ` argument here could be left implicit, as well as in all the lemmas in `NatSolver.EvaluationHomomorphism`. Doesn't really matter, but it would make the code more readable (esp in the latter case). I guess the same change is possible for the ring solver?

Reply:
Ohno. Why didn't I do this?
Thanks - it works for the NatSolver, I'll do the same for the RingSolver now.

Reply:
Also works for the RingSolver